pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go

```diff
@@ -0,0 +1,218 @@
+		return nil, fmt.Errorf("ClusterTrustBundle informer has not yet synced")
+	}
+
+	ctb, err := m.ctbLister.Get(name)
+	if k8serrors.IsNotFound(err) && allowMissing {
+		return []byte{}, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("while getting ClusterTrustBundle: %w", err)
+	}
+
+	cacheKey := cacheKeyType{ctbName: name}
+
+	if cachedAnchors, ok := m.normalizationCache.Get(cacheKey); ok {
+		return cachedAnchors.([]byte), nil
+	}
+
+	pemTrustAnchors, err := m.normalizeTrustAnchors([]*certificatesv1alpha1.ClusterTrustBundle{ctb})
+	if err != nil {
+		return nil, fmt.Errorf("while normalizing trust anchors: %w", err)
+	}
+
+	m.normalizationCache.Add(cacheKey, pemTrustAnchors, 5*time.Minute)
+
+	return pemTrustAnchors, nil
+}
+
+func (m *InformerManager) GetTrustAnchorsBySigner(signerName string, labelSelector *metav1.LabelSelector, allowMissing bool) ([]byte, error) {
+	if !m.ctbInformer.HasSynced() {
+		return nil, fmt.Errorf("ClusterTrustBundle informer has not yet synced")
+	}
+
+	// Note that this function treats nil as "match nothing", and non-nil but
+	// empty as "match everything".
+	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("while parsing label selector: %w", err)
+	}
+
+	rawCTBList, err := m.ctbLister.List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("while listing ClusterTrustBundles matching label selector %v: %w", labelSelector, err)
+	}
+
+	ctbList := []*certificatesv1alpha1.ClusterTrustBundle{}
+	for _, ctb := range rawCTBList {
+		if ctb.Spec.SignerName == signerName {
+			ctbList = append(ctbList, ctb)
+		}
+	}
+
+	if len(ctbList) == 0 {
+		if allowMissing {
+			return []byte{}, nil
+		}
+		return nil, fmt.Errorf("combination of signerName and labelSelector matched zero ClusterTrustBundles")
+	}
+
+	cacheKey := cacheKeyType{signerName: signerName, labelSelector: selector.String()}
+
+	if cachedAnchors, ok := m.normalizationCache.Get(cacheKey); ok {
+		return cachedAnchors.([]byte), nil
+	}
+
+	pemTrustAnchors, err := m.normalizeTrustAnchors(ctbList)
+	if err != nil {
+		return nil, fmt.Errorf("while normalizing trust anchors: %w", err)
+	}
+
+	m.normalizationCache.Add(cacheKey, pemTrustAnchors, 5*time.Minute)
+
+	return pemTrustAnchors, nil
+}
+
+func (m *InformerManager) normalizeTrustAnchors(ctbList []*certificatesv1alpha1.ClusterTrustBundle) ([]byte, error) {
+	// Deduplicate trust anchors from all ClusterTrustBundles that match signerName and labelSelector.
+	trustAnchorSet := map[string]bool{}
+	for _, ctb := range ctbList {
+		rest := []byte(ctb.Spec.TrustBundle)
+		var b *pem.Block
+		for {
+			b, rest = pem.Decode(rest)
+			if b == nil {
+				break
+			}
+			trustAnchorSet[string(b.Bytes)] = true
+		}
+	}
+
+	trustAnchors := make([]string, 0, len(trustAnchorSet))
+	for ta := range trustAnchorSet {
+		trustAnchors = append(trustAnchors, ta)
+	}
+	sort.Strings(trustAnchors)
+
+	// Reserialize the deduped and sorted set to PEM.
```

Comment:
Uh I thought we wanted the ordering of the PEMs to be random every time?

Reply:
Not random --- we don't want to spam the client with changed file content.  We just want to prevent user from being able to rely on the ordering of the certificates in the ClusterTrustBundle object.

Reply:
Right I mean "generate a seed on kubelet start up and use that to sort the list in a deterministic but random way" (there are other ways to go about it as well)